### ISSUE TITLE
Remove code constructs to support legacy Python 2 and Python 3 versions

### ISSUE THREAD

**Issue by MattHag** [pwr-Solaar/Solaar#2273]
**Information**

- Solaar version: 1.11rc (latest master)

**Is your feature request related to a problem? Please describe.**
Although, setup.py mentions compatibility for Python 3.7 and above, there are still some additional code pieces to support Python 2. A cleanup should remove these workarounds.

**Comment by pfps** [comment]
Yeah, the question is identifying what should be removed.

**Comment by MattHag** [comment]
At least the code pieces commented with Python 2, which I already removed and pushed. 

Maybe 'Upgrade code to Python 3 style' is a better title, in order to upgrade strings to f-strings with pyupgrade as well and upgrade remaining strings manually.
```
pip install pyupgrade
pyupgrade `find . -name "*.py" -type f`
pyupgrade `find . -name "*.py" -type f`
```

**Comment by MattHag** [comment]
In addition to that, I also suggest a switch to double-quoted strings "" instead of '', and an upgrade to the insanely fast `ruff`, which replaces isort, yapf and flake8. This tool is also quite handy as very quick' on save' auto formatter.

**Comment by pfps** [comment]
These seem reasonable, but let's defer this until after 1.11.1 goes out, which should be in a week.   Then I have a bunch of changes to improve the imports loops, except for the big one of calling the UI from logitech_receiver.

**Comment by pfps** [comment]
The advice I have received is that the license notice should be at the top of each file.  It's probably overkill but that is what I have been told.

**Comment by MattHag** [comment]
Yes, it is and it is not beneficial for development and leads to some irritation for me.

It is obvious, that there's no real plan for maintaining them, since most of the years data back to 2012-2013. This needs a plan and the best is to remove unnecessary duplication and ambiguous years from the individual files.

- Which name should be there?
  - If it is Solaar developers, is that even something good or does that refer to anyone contributing something. Could this have a weakening effect on the legal side?
- What does the year refer to? 2013 is certainly not the year of the latest changes, but also not the introduction of the module.

**Comment by pfps** [comment]
The important part of this section is the license.   The code is Solaar uses the GPL and this can't be changed and should be prominently mentioned in each file.

It is true that the copyright line has not been updated.  I'm now putting new information in for the files that I am touching.

**Comment by MattHag** [comment]
Good, then a cleanup and unification would be to replace all existing ones with that timeless message, as 90% of them are dated.
And furthermore reformat the module level descriptions, which are somtetimes appended in the same style as the copyright notice. They use a # instead of a Multiline comment „““, which is easy to notice.

**Comment by MattHag** [comment]
> It is true that the copyright line has not been updated. I'm now putting new information in for the files that I am touching.

This needs a pre-commit hook then, otherwise it will never be up-to-date.

**Comment by pfps** [comment]
The copyright lines are the least important part of the license block, or so I understand.  Having them be out of date is not ideal but doesn't invalidate the license.

**Comment by MattHag** [comment]
Sure, but there's certainly a nice handy tool to maintain it. Or, I would not even add a date at all.

**Comment by MattHag** [comment]
Remove humongous number of code lines in an __init__.py.

**Comment by rloutrel** [comment]
Hi. I wish I could help for this task, since it is a good way to discover the project and step up.

I am a pretty good soldier, but I do not know the battlefield and the HQ-objectives, so I should be instructed.

I was thinking, that the point:
- [ ] Remove all assertion from non-test code

Would be feasible without a deep knowledge of the project.

I understand, that the point is to replace with "try ... except ... " the "assert" assertions in the productive code, since it is meant to be testing purpose. Is this correct?
Could i have a simple example of the expectations before / after ?

My dev/test environments will be:
* OS: Fedora Plasma (KDE) | Ubuntu Gnome
* Mouses: MX Anywhere 2S / MX Anywhere 3 / Mx Master
* Keyboards: MX Keys / Mx 400+ (for the record: the white one)
* No copilot, so I will fight alone on this

**Comment by rloutrel** [comment]
Ok. Then I will do the [NamedInt, NamedInts, UnsortedNamedInts] part.

About the tests, I have 28 from the 470 tests that are failing. Is this something, that I can first ignore (some of the tests seems quite recent and maybe not stable yet / in development)

**Comment by rloutrel** [comment]
I am thinking about starting with the class _SettingWithValueUI from diversion_rules.py and the list of settings.

But I am a bit clueless about the tests to create. I will think about it.

**Comment by MattHag** [comment]
The tree model stuff likely needs to change for GTK 4.0, so I'd not touch it right now. 

> About the tests, I have 28 from the 470 tests that are failing. Is this something, that I can first ignore (some of the tests seems quite recent and maybe not stable yet / in development)

Starting from my latest pull request #2626 all tests should pass. I think the hidpp constants contain many NamedInts, which seem doable to replace, especially the ones without fallback value. Some modules like diversion/rules at not teddy for testing yet, but if a function is well testable on its own a test helps.

**Comment by rloutrel** [comment]
I saw the diff of your PR. I will wait for the merge before any modifications :smiley: 

I can not work on it before this weekend.

Do you have WIP on hidapi (I saw #2480 is in progress), which would conflict with hidpp (not sure about it, I did not check the code about this part): to avoid conflicts, I would use your branch.

**Comment by rloutrel** [comment]
Is this:
https://github.com/pwr-Solaar/Solaar/compare/master...rloutrel:Solaar:master

What you are expecting for the testing, before I start with the refactoring?

**Comment by MattHag** [comment]
Yes that looks good

**Comment by WhyNotHugo** [comment]
setup.py is generally deprecated in the Python ecosystem, but all the new fancy build systems don't support installing files outside the python module (e.g.: desktop entry, icon, metainfo XML).

I think that meson is the best fit for solaar, I can draft a PR moving to meson and tidying up the build system is you're interested.

**Comment by WhyNotHugo** [comment]
I also think it's reasonable to drop support for Python < 3.10; these versions are no longer supported by upstream.
